dataflow/gemma-flex-template/README.md

```diff
@@ -0,0 +1,179 @@
+          checkpoint_path: the path to a local copy of gemma model weights.
+          tokenizer_path: the path to a local copy of the gemma tokenizer
+          device: optional. the device to run inference on. can be either
+            'cpu' or 'gpu', defaults to cpu. 
+        """
+        model_config = get_config_for_2b(
+        ) if "2b" in model_variant else get_config_for_7b()
+        model_config.tokenizer = tokenizer_path
+        model_config.quant = 'quant' in model_variant
+        model_config.tokenizer = tokenizer_path
+
+        self._model_config = model_config
+        self._checkpoint_path = checkpoint_path
+        if device == 'GPU':
+            logging.info("Device is set to CUDA")
+            self._device = torch.device('cuda')
+        else:
+            logging.info("Device is set to CPU")
+            self._device = torch.device('cpu')
+        self._env_vars = {}
+
+    def load_model(self) -> GemmaForCausalLM:
+        """Loads and initializes a model for processing."""
+        torch.set_default_dtype(self._model_config.get_dtype())
+        model = GemmaForCausalLM(self._model_config)
+        model.load_weights(self._checkpoint_path)
+        model = model.to(self._device).eval()
+        return model
+```
+
+### Formatting DoFn
+
+The output from a keyed model handler is a tuple of the form `(key, PredictionResult)`. To format that output into a string before sending it to the answer Pub/Sub topic, use an extra `DoFn`.
+
+```py
+class FormatOutput(beam.DoFn):
+  def process(self, element, *args, **kwargs):
+    yield "Key : {key}, Input: {input}, Output: {output}".format(key=element[0], input=element[1].example, output=element[1].inference)
+```
+
+## Build the Flex Template
+Run the following code from the directory to build the Dataflow flex template.
+
+- Replace `$TEMPLATE_FILE` with a Google Cloud Storage path written as a JSON file.
+- Set `SDK_CONTAINER_IMAGE` to the name of the Docker image created previously.
+- `$PROJECT` is the Google Cloud project that you created previously. 
+
+```sh
+gcloud dataflow flex-template build $TEMPLATE_FILE \
+  --image $SDK_CONTAINER_IMAGE \
+  --sdk-language "PYTHON" \
+  --metadata-file metadata.json \
+  --project $PROJECT
+```
+
+## Start the pipeline
+To start the Dataflow streaming job, run the following code from the directory. Replace `$TEMPLATE_FILE`, `$REGION`, `$GCS_BUCKET`, `$INPUT_SUBSCRIPTION`, `$OUTPUT_TOPIC`, `$SDK_CONTAINER_IMAGE`, and `$PROJECT` with the Google Cloud project resources you created previously. It might take as much as 15 minutes for the worker to start up and to begin accepting messages from the input Pub/Sub topic. 
+
+```sh
+gcloud dataflow flex-template run "gemma-flex-`date +%Y%m%d-%H%M%S`" \
+  --template-file-gcs-location $TEMPLATE_FILE \
+  --region $REGION \
+  --staging-location $GCS_BUCKET \
+  --parameters messages_subscription=$INPUT_SUBSCRIPTION \
+  --parameters responses_topic=$OUTPUT_TOPIC \
```

Comment:
These need to be the fully qualified pubsub subscription and pubsub topic name. I received errors trying to use just the topic/subscription name.

Reply:
clarified